### PR TITLE
Refactor format to accommodate additional formatting steps

### PR DESCRIPTION
**Pull request by JoeRobich**
Format was built around running only the whitespace formatting provided by Roslyn's Formatter. This goal of this PR is to refactor to allow for additional formatting steps to be added.

- Separate out determining which files to consider formattable
- Separate out applying formatting to the formattable documents

### CODE REVIEW

File: src/Formatters/ICodeFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.CodeAnalysis.Tools.Formatters
+{
+    internal interface ICodeFormatter
+    {
+        Task<Solution> FormatAsync(ILogger logger, Solution solution, ImmutableArray<(Document, OptionSet)> formatableDocuments, CancellationToken cancellationToken);
```
Comment (jmarolf) [inline src/Formatters/ICodeFormatter.cs]: I would add the logger after `formatableDocuments` but before `cancellationToken`

File: src/Formatters/AbstractCodeFormatter.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            var formattedDocuments = FormatFiles(logger, formatableDocuments, cancellationToken);
+            return await ApplyFileChangesAsync(logger, solution, formattedDocuments);
+        }
+
+        protected abstract ImmutableArray<(Document, Task<SourceText>)> FormatFiles(
+            ILogger logger, 
+            ImmutableArray<(Document, OptionSet)> formatableDocuments, 
+            CancellationToken cancellationToken);
+
+        private static async Task<Solution> ApplyFileChangesAsync(
```
Comment (jmarolf) [inline src/Formatters/AbstractCodeFormatter.cs]: I assume the purpose of this method will be made clear in the fullness of time

Comment (JoeRobich) [reply src/Formatters/AbstractCodeFormatter.cs]: I'll add comments. =)

File: tests/dotnet-format.UnitTests.csproj [hunk 3]
```diff
@@ -7,8 +7,11 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <Compile Remove="obj\**" />
     <Compile Remove="projects\**" />
+    <EmbeddedResource Remove="obj\**" />
     <EmbeddedResource Remove="projects\**" />
+    <None Remove="obj\**" />
```
Comment (sharwell) [inline tests/dotnet-format.UnitTests.csproj]: ❗️  These lines should not have been added.

File: src/CodeFormatter.cs [hunk 4]
```diff
@@ -22,19 +21,106 @@ namespace Microsoft.CodeAnalysis.Tools
...
                 ExitCode = 1
             };
-            var workspaceStopwatch = Stopwatch.StartNew();
+
+            using (var workspace = await OpenWorkspaceAsync(
+                solutionOrProjectPath, isSolution, logAllWorkspaceWarnings, logger, cancellationToken).ConfigureAwait(false))
+            {
+                if (workspace is null)
+                {
+                    return formatResult;
```
Comment (jmarolf) [inline src/CodeFormatter.cs]: Perhaps have `OpenWorkspaceAsync` return a `WorkspaceFormatResult`?

**Comment by JoeRobich** [comment]
@sharwell could you take another look?

**Pull request merged**